### PR TITLE
Feature/zonedto

### PR DESCRIPTION
**Pull request by aliasgharmhowwala**
New Request with Escape character fix

### CODE REVIEW

File: zone.go [hunk 1]
```diff
@@ -50,7 +50,8 @@ type ZoneKey struct {
 
 // URI generates the URI for an Zone
 func (k ZoneKey) URI() string {
-	uri := fmt.Sprintf("zones/?&q=name:%s", k.Zone)
+	zoneName := strings.Replace(k.Zone, "/", "%2F", -1)
```
Comment (kbhandari) [inline zone.go:54]: Please add code comment

File: udnssdk_test.go [hunk 2]
```diff
@@ -215,3 +216,26 @@ func Test_CheckResponse_StatusCode4xx(t *testing.T) {
...
+// Testcase for checking up the escaping character "/"
+func Test_EscapeCharacter_Zone(t *testing.T) {
+
+	testClient, _ := NewClient(testUsername, testPassword, testBaseURL)
+	req, _ := testClient.NewRequest("GET", "https://api.test.ultradns.net/zones/0%2F1.70.78.208.in-addr.arpa./rrsets/ANY?&offset=0&limit=1000", RRSetListDTO{})
+	assert.Equal(t, req.URL.RawPath, "/v1/https://api.test.ultradns.net/zones/0%2F1.70.78.208.in-addr.arpa./rrsets/ANY")
+
+}
+
+func Test_CustomHeader(t *testing.T){
```
Comment (kbhandari) [inline udnssdk_test.go:230]: Please add a test for the scenario when custom header is not set.

File: udnssdk.go [hunk 3]
```diff
@@ -148,14 +156,20 @@ func (c *Client) NewRequest(method, pathquery string, payload interface{}) (*htt
 		}
 	}
 
-	req, err := http.NewRequest(method, url.String(), body)
+	urlString := strings.Replace(url.String(), "%252F", "%2F", -1)
```
Comment (kbhandari) [inline udnssdk.go:157]: Please add code comment

File: rrset.go [hunk 4]
```diff
@@ -263,11 +264,13 @@ type RRSetKey struct {
 
 // URI generates the URI for an RRSet
 func (k RRSetKey) URI() string {
-	uri := fmt.Sprintf("zones/%s/rrsets", k.Zone)
+	zoneName := strings.Replace(k.Zone, "/", "%2F", -1)
```
Comment (kbhandari) [inline rrset.go:268]: Please add code comment

File: rrset.go [hunk 5]
```diff
@@ -263,11 +264,13 @@ type RRSetKey struct {
...
+	uri := fmt.Sprintf("zones/%s/rrsets", zoneName)
 	if k.Type != "" {
-		uri += fmt.Sprintf("/%v", k.Type)
+		uri += fmt.Sprintf("/%s", k.Type)
 		if k.Name != "" {
-			uri += fmt.Sprintf("/%v", k.Name)
+			ownerName := strings.Replace(k.Name, "/", "%2F", -1)
```
Comment (kbhandari) [inline rrset.go:274]: Please add code comment

File: udnssdk.go [hunk 6]
```diff
@@ -32,6 +32,14 @@ const (
 	apiVersion = "v1"
```
Comment (kbhandari) [inline udnssdk.go]: We don't need to append version anymore to the uri.

File: zone.go (hunk 1 above)
Comment (aliasgharmhowwala) [reply zone.go:54]: Added code comment

File: zone_test.go [hunk 7]
```diff
@@ -59,11 +59,9 @@ func Test_Zone_SelectWithOffsetWithLimit_WithZoneKey(t *testing.T) {
 		}
 		offset = ri.ReturnedCount + ri.Offset
 		continue
-		if err != nil {
-			t.Fatal(err)
-		}
 
 	}
+	assert.Equal(t,zones[0].Properties.Name,testDomain)
```
Comment (aliasgharmhowwala) [inline zone_test.go:64]: Added assertions

File: zone_test.go [hunk 8]
```diff
@@ -112,10 +110,7 @@ func Test_Zone_SelectWithOffsetWithLimit_WithOutAnyValue(t *testing.T) {
 		offset = ri.ReturnedCount + ri.Offset
 		continue
 	}
-
-	if err != nil {
-		t.Fatal(err)
-	}
+	assert.NotNil(t,zones)
```
Comment (aliasgharmhowwala) [inline zone_test.go:113]: added assertion for not null

File: zone_test.go [hunk 9]
```diff
@@ -136,29 +131,10 @@ func Test_Zone_InvalidZone(t *testing.T) {
...
-			}
-			t.Logf("%v", err)
-			return
-		} else {
-			t.Fatal(" Expected to fail")
-
-		}
-	}
+	_, _, _, err = testClient.Zone.SelectWithOffsetWithLimit(r, offset, limit)
+	assert.NotNil(t,err)
```
Comment (aliasgharmhowwala) [inline zone_test.go:137]: enhanced testcase with assertions

File: zone_test.go [hunk 10]
```diff
@@ -179,29 +155,12 @@ func Test_Zone_InvalidAccount(t *testing.T) {
...
-					time.Sleep(waittime)
-					continue
-				}
-			}
-			t.Logf("%v", err)
-			return
-		} else {
-			t.Fatal(" Expected to fail")
+	_, _, _, err = testClient.Zone.SelectWithOffsetWithLimit(r, offset, limit)
+        assert.NotNil(t,err)
```
Comment (aliasgharmhowwala) [inline zone_test.go:162]: enhanced testcases with assertions

File: udnssdk_test.go [hunk 11]
```diff
@@ -249,5 +249,5 @@ func Test_EmptyCustomHeader(t *testing.T){
```
Comment (aliasgharmhowwala) [inline udnssdk_test.go:249]: New testcase to check empty header

**Review by kbhandari — APPROVED** [review]

**Pull request merged**